executor/aggregate.go

```diff
@@ -45,39 +96,482 @@ type HashAggExec struct {
+				for i, af := range w.aggFuncs {
+					if err = af.Update(aggEvalCtxs[i], sc, row); err != nil {
+						return errors.Trace(err)
+					}
+				}
+			}
+			w.intermDataRowsBuffer = w.intermDataRowsBuffer[:0]
+			if !ok {
+				return
+			}
+		}
+		var reachEnd bool
+		if w.intermDataRowsBuffer, reachEnd = input.ToRows(sc, w.intermDataRowsBuffer, w.aggFuncs, w.maxChunkSize); reachEnd {
+			input = nil
+		}
+	}
+	return
+}
+
+func (w *HashAggFinalWorker) getFinalResult(sc *stmtctx.StatementContext) {
+	groupIter := w.groupSet.NewIterator()
+	result, ok := <-w.finalResultHolderCh
+	if !ok {
+		return
+	}
+	result.Reset()
+	for {
+		groupKey, _ := groupIter.Next()
+		if groupKey == nil {
+			if result.NumRows() > 0 {
+				w.outputCh <- &AfFinalResult{chk: result, giveBackCh: w.finalResultHolderCh}
+			}
+			return
+		}
+		aggCtxs := w.getContext(sc, groupKey, w.aggCtxsMap)
+		w.rowBuffer = w.rowBuffer[:0]
+		for i, af := range w.aggFuncs {
+			w.rowBuffer = append(w.rowBuffer, af.GetResult(aggCtxs[i]))
+		}
+		w.mutableRow.SetDatums(w.rowBuffer...)
+		result.AppendRow(w.mutableRow.ToRow())
+		if result.NumRows() == w.maxChunkSize {
+			w.outputCh <- &AfFinalResult{chk: result, giveBackCh: w.finalResultHolderCh}
+			result, ok = <-w.finalResultHolderCh
+			if !ok {
+				return
+			}
+			result.Reset()
+		}
+	}
+}
+
+func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		waitGroup.Done()
+	}()
+	sc := ctx.GetSessionVars().StmtCtx
+	if err := w.fetchIntermData(sc); err != nil {
+		w.outputCh <- &AfFinalResult{err: errors.Trace(err)}
+	}
+	w.getFinalResult(sc)
+}
+
 // Next implements the Executor Next interface.
 func (e *HashAggExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.doesUnparallelExec {
+		return errors.Trace(e.unparallelExec(ctx, chk))
+	}
+	return errors.Trace(e.parallelExec(ctx, chk))
+}
+
+func (e *HashAggExec) fetchChildData(ctx context.Context) {
+	var (
+		input *HashAggInput
+		chk   *chunk.Chunk
+		ok    bool
+		err   error
+	)
+	defer func() {
+		for _, w := range e.partialWorkers {
+			close(w.inputCh)
+		}
+	}()
+	for {
+		select {
+		case <-e.finishCh:
+			return
+		case input, ok = <-e.inputCh:
+			if !ok {
+				return
+			}
+			chk = input.chk
+		}
+		err = e.children[0].Next(ctx, chk)
+		if err != nil {
+			e.finalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		if chk.NumRows() == 0 {
+			return
+		}
+		input.giveBackCh <- chk
+	}
+}
+
+func (e *HashAggExec) waitPartialWorkerAndCloseOutputChs(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	for _, ch := range e.partialOutputChs {
+		close(ch)
+	}
+}
+
+func (e *HashAggExec) waitFinalWorkerAndCloseFinalOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	close(e.finalOutputCh)
+}
+
+// parallelExec executes hash aggregate algorithm parallelly.
+func (e *HashAggExec) parallelExec(ctx context.Context, chk *chunk.Chunk) error {
+	if !e.prepared {
+		go e.fetchChildData(ctx)
```

Comment:
Could you extract a function like `prepareParallelExeciton` to do these goroutine start works:
```go
if !e.prepared {
    e.prepareParallelExecution()
    e.prepared = true
}
```